Addition of OwnerData and OptionalCompanyData to the output of bills/getAll (#378)
MewsSystems/gitbook-connector-api: operations/bills.md

            },
            "OrderItems": [],
            "PaymentItems": [],
            "OwnerData": {

>                 "ID": "26afba60-06c3-455b-92db-0e3983be0b1d",

Reviewer: The ID will be inside the value, right?

            },
            "OrderItems": [],
            "PaymentItems": [],
            "OwnerData": {

>                 "discriminator": "BillCustomerData",

Reviewer 2: ```suggestion
                "Discriminator": "BillCustomerData",
```

>                 "value": {

Reviewer 2: ```suggestion
                "Value": {
```

>                     "ID": "26afba60-06c3-455b-92db-0e3983be0b1d",

Reviewer 2: ```suggestion
                    "Id": "26afba60-06c3-455b-92db-0e3983be0b1d",
```

                    "Address": {
                        "Line1": "Joe Doe street",
                        "Line2": "Very long ave",
                        "City": "Townston",
                        "PostalCode": "154 00",
                        "SubdivisionCode": "AU-NSW",
                        "CountryCode": "AU"
                    },
                    "LegalIdentifiers": {
                        "LegalIdentifier": "Value of legal identifier",
                        "LegalIdentifier2": "Value of legal identifier 2",
                    },
                    "BillingCode": "billing code value",
                    "LastName": "Last name value",
                    "FirstName": "First name value",
                    "SecondLastName": "Second last name value",
                    "TitlePrefix": "Mistress",
                    "FiscalIdentifier": "Fiscal identifier",

>                     "AdditionalTaxIdentifier": "Additional tax identifier"

Reviewer 3: Can we put here more real data like in the Address please?

Author: I did what I could :) couldn't find more info on the other fields.

                }
            },
            "CompanyDetails": {
                "Id": "26afba60-06c3-455b-92db-0e3983be0b1d",
                "Address": {
                    "Line1": "Joe Doe street",
                    "Line2": "Very long ave",
                    "City": "Townston",
                    "PostalCode": "154 00",
                    "SubdivisionCode": "AU-NSW",
                    "CountryCode": "AU"
                },
                "LegalIdentifiers": {
                    "LegalIdentifier": "Value of legal identifier",
                    "LegalIdentifier2": "Value of legal identifier 2",
                },
                "BillingCode": "billing code value",
                "Name": "name value",
                "FiscalIdentifier": "Fiscal identifier",
                "AdditionalTaxIdentifier": "Additional tax identifier"

>             },

Reviewer 3: Can we put here more real data like in the Address please?

            },
            "OrderItems": [],
            "PaymentItems": [],
            "OwnerData": {
                "Discriminator": "BillCustomerData",
                "Value": {
                    "Id": "26afba60-06c3-455b-92db-0e3983be0b1d",
                    "Address": {
                        "Line1": "Joe Doe street",
                        "Line2": "Very long ave",
                        "City": "Townston",
                        "PostalCode": "154 00",
                        "SubdivisionCode": "AU-NSW",
                        "CountryCode": "AU"
                    },
                    "LegalIdentifiers": {

>                         "LegalIdentifier": "Value of legal identifier",

Reviewer 2: Please use real examples here

                        "LegalIdentifier2": "Value of legal identifier 2",
                    },
                    "BillingCode": "Billing code value",
                    "LastName": "Doe",
                    "FirstName": "John",
                    "SecondLastName": "Vincent",
                    "TitlePrefix": "Mistress",
                    "FiscalIdentifier": "Fiscal identifier",
                    "AdditionalTaxIdentifier": "Additional tax identifier"
                }
            },
            "CompanyDetails": {
                "Id": "26afba60-06c3-455b-92db-0e3983be0b1d",
                "Address": {
                    "Line1": "Joe Doe street",
                    "Line2": "Very long ave",
                    "City": "Townston",
                    "PostalCode": "154 00",
                    "SubdivisionCode": "AU-NSW",
                    "CountryCode": "AU"
                },

>                 "LegalIdentifiers": {

Reviewer 2: Please use real examples here

                    "ItalianFiscalCode": null
                }
            },

>             "AssigneeData": {

Reviewer: Yes, let's remove AssigneeData.

                "discriminator": "BillCustomerData",
                "value": {

>                 "Address": {

Reviewer 2: The indenting doesn't seem right here, can you review it?

                    "Line1": "Joe Doe street",
                    "Line2": "Very long ave",
                    "City": "Townston",
                    "PostalCode": "154 00",
                    "SubdivisionCode": "AU-NSW",
                    "CountryCode": "AU"
                },
                "LegalIdentifiers": {
                    "LegalIdentifier": "ValueOfLegalIdentifier",
                },
                "BillingCode": "billing code value",
                "Name": "name value",
                "FiscalIdentifier": "Fiscal identifier",
                "AdditionalTaxIdentifier": "Additional tax identifier"
                }
            },
            "OwnerData": {
                "discriminator": "BillCustomerData",
                "value": {

>                     "Address": {

Reviewer: ID

                        "Line1": "Joe Doe street",
                        "Line2": "Very long ave",
                        "City": "Townston",
                        "PostalCode": "154 00",
                        "SubdivisionCode": "AU-NSW",
                        "CountryCode": "AU"
                    },
                    "LegalIdentifiers": {
                        "LegalIdentifier": "Value of legal identifier",
                        "LegalIdentifier2": "Value of legal identifier 2",
                    },
                    "BillingCode": "billing code value",
                    "Name": "name value",
                    "FiscalIdentifier": "Fiscal identifier",
                    "AdditionalTaxIdentifier": "Additional tax identifier"
                }
            },

>             "OptionalCompanyData": {

Reviewer: ```suggestion
            "CompanyDetails": {
```

| `TrackReceivable` | boolean | required | Tracking of payments is enabled for bill, only applicable for `Invoice`. |
| `DisplayCid` | boolean | required | Display CID number on bill, only applicable for `Invoice`. |

#### Bill owner data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Discriminator` | string [Bill owner data discriminator](#bill-owner-data-discriminator) | required | Determines type of value. |

> | `Value` | object | required | Structure of object depends on [Bill owner data discriminator](#bill-owner-data-discriminator). |

Reviewer: We should mention the 2 cases available and link to those.

| `TrackReceivable` | boolean | required | Tracking of payments is enabled for bill, only applicable for `Invoice`. |
| `DisplayCid` | boolean | required | Display CID number on bill, only applicable for `Invoice`. |

#### Bill owner data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Discriminator` | string [Bill owner data discriminator](#bill-owner-data-discriminator) | required | Determines type of value. |
| `Value` | object | required | Structure of object depends on [Bill owner data discriminator](#bill-owner-data-discriminator). Can be either of type [Bill customer data](#bill-customer-data) or [Bill company data](#bill-company-data)|

#### Bill owner data discriminator

> * `BillCustomerData` - Assignee data specific to a customer.

Reviewer 2: ```suggestion
* `BillCustomerData` - Owner data specific to a customer.
```

> * `BillCompanyData` - Assignee data specific to a company.

Reviewer 2: ```suggestion
* `BillCompanyData` - Owner data specific to a company.
```

| `TrackReceivable` | boolean | required | Tracking of payments is enabled for bill, only applicable for `Invoice`. |
| `DisplayCid` | boolean | required | Display CID number on bill, only applicable for `Invoice`. |

#### Bill owner data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Discriminator` | string [Bill owner data discriminator](#bill-owner-data-discriminator) | required | Determines type of value. |

> | `Value` | object | required | Structure of object depends on [Bill owner data discriminator](#bill-owner-data-discriminator). Can be either of type [Bill customer data](#bill-customer-data) or [Bill company data](#bill-company-data)|

Reviewer 3: ```suggestion
| `Value` | object | required | Structure of object depends on [Bill owner data discriminator](#bill-owner-data-discriminator). Can be either of type [Bill customer data](#bill-customer-data) or [Bill company data](#bill-company-data). |
```

#### Bill owner data discriminator

> * `BillCustomerData` - Owner data specific to a customer.
> * `BillCompanyData` - Owner data specific to a company.

Reviewer 3: Small improvement imho

```suggestion
* `BillCustomerData` - Owner data specific to a [Customer](customers.md#customer).
* `BillCompanyData` - Owner data specific to a [Company](companies.md#company).
```

| `LegalIdentifiers` | [LegalIdentifiers](#legal-identifiers) | optional | Legal Identifiers of the customer. |
| `BillingCode` | string  | optional | A unique code for Mews to list on invoices it sends to the customer. |
| `LastName` | string  | optional | Last name of the customer. |
| `FirstName` | string  | optional | First name of the customer. |
| `SecondLastName` | string  | optional | Second last name of the customer. |
| `TitlePrefix` | [Title](customers.md#Title)  | optional | Title prefix of the customer. |

#### Bill company data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Id` | string  | optional | ID of the [Company](companies.md#company). |
| `Address` | [Bill address](#bill-address)  | optional | Address of the company. |
| `LegalIdentifiers` | [LegalIdentifiers](#legal-identifiers) | optional | Legal Identifiers of the company. |
| `BillingCode` | string  | optional | A unique code for Mews to list on invoices it sends to the company. |
| `Name` | string  | optional | Name of the company. |
| `FiscalIdentifier` | string  | optional | Fiscal identifier of the company. |

> | `AdditionalTaxIdentifier` | string  | optional | Additional tax identifier of the company. |

Reviewer 3: Can you double check `Contract` here please? I don't think all of them are optional.
There is a discriminator which tells you if there is a customer or company but this doesn't relate to the Contract here. This contract column determinates only if the entity is provided so I think of example `Id` is always required.

Same goes for the customer please.

I hope my poor explanation is good 😅

| `OrderItems` | array of [Order item](accountingitems.md#order-item) | required | The order items (consumed items such as nights or products) on the bill. |
| `PaymentItems` | array of [Payment item](accountingitems.md#payment-item) | required | The payment items (such as cash, credit card payments or invoices) on the bill. |
| `AssigneeData` | [Bill assignee data](#bill-assignee-data) | optional | Additional information about assignee of the bill. Persisted at the time of closing of the bill. |

> | `OwnerData` | [Bill owner data](#bill-assignee-data) | optional | Additional information about assignee of the bill. Persisted at the time of closing of the bill. |

Reviewer: With references.
```suggestion
| `OwnerData` | [Bill owner data](#bill-assignee-data) | optional | Additional information about owner of the bill. Can be a Customer or Company. Persisted at the time of closing of the bill. |
```

> | `OptionalCompanyData` | [Bill company data](#bill-company-data) | optional | Additional information about company assigned to the bill. Persisted at the time of closing of the bill. |

Reviewer: ```suggestion
| `OptionalCompanyData` | [Bill company data](#bill-company-data) | optional | Additional information about the company assigned to the bill. Not the same as the owner. Persisted at the time of closing of the bill. |
```

| `TrackReceivable` | boolean | required | Tracking of payments is enabled for bill, only applicable for `Invoice`. |
| `DisplayCid` | boolean | required | Display CID number on bill, only applicable for `Invoice`. |

#### Bill assignee data
#### Bill owner data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `Discriminator` | string [Bill assignee data discriminator](#bill-assignee-data-discriminator) | required | Determines type of value. |

Author: We've decided to drop the assignee naming. This is just a rename to make the description more in sync with company billing.

#### Bill owner data discriminator

* `BillCustomerData` - Owner data specific to a [Customer](customers.md#customer).
* `BillCompanyData` - Owner data specific to a [Company](companies.md#company).

#### Bill customer data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Id` | string  | required | ID of the [Customer](customers.md#customer) to whom the bill was assigned. |
| `Address` | [Bill address](#bill-address) | optional | Address of the customer. |

> | `LegalIdentifiers` | [LegalIdentifiers](#legal-identifiers) | optional | Legal Identifiers of the customer. |

Reviewer 2: ```suggestion
| `LegalIdentifiers` | [Dictionary](#Dictionary) | optional | The set of [legal identifiers](#legal-identifiers) for the company. |
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `CustomerId` | string  | optional | ID of the [Customer](customers.md#customer) to whom the bill was assigned. |

Reviewer: ```suggestion
| `Id` | string  | optional | ID of the [Customer](customers.md#customer) to whom the bill was assigned. |
```

| `Address` | [Address](#address)  | optional | Address of the customer. |
| `LegalIdentifiers` | Dictionary<string, string>  | optional | Legal Identifiers of the customer. |
| `BillingCode` | string  | optional | ID of the Customer to whom the bill was assigned. |
| `LastName` | string  | optional | Last name of the customer. |
| `FirstName` | string  | optional | First name of the customer. |
| `SecondLastName` | string  | optional | Second last name of the customer. |
| `TitlePrefix` | string  | optional | Title prefix of the customer. |

#### Bill company data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `CompanyId` | string  | optional | ID of the [Company](companies.md#company) to whom the bill was assigned. |

Reviewer: It needs to match the response
```suggestion
| `Id` | string  | optional | ID of the [Company](companies.md#company) to whom the bill was assigned. |
```

> | `Address` | [Bill address](#bill-address)  | optional | Address of the customer. |
> | `LegalIdentifiers` | Dictionary<string, string>  | optional | Legal Identifiers of the customer. |
> | `BillingCode` | string  | optional | ID of the Customer to whom the bill was assigned. |
> | `Name` | string  | optional | Name of the company to which the bill was assigned. |
> | `FiscalIdentifier` | string  | optional | Fiscal identifier of the company. |
> | `AdditionalTaxIdentifier` | string  | optional | Additional tax identifier of the company. |

Reviewer: ```suggestion
| `Address` | [Bill address](#bill-address)  | optional | Address of the company. |
| `LegalIdentifiers` | Dictionary<string, string>  | optional | Legal Identifiers of the company. |
| `BillingCode` | string  | optional | ID of the company to whom the bill was assigned. |
| `Name` | string  | optional | Name of the company to which the bill was assigned. |
| `FiscalIdentifier` | string  | optional | Fiscal identifier of the company. |
| `AdditionalTaxIdentifier` | string  | optional | Additional tax identifier of the company. |
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Id` | string  | optional | ID of the [Customer](customers.md#customer) to whom the bill was assigned. |

> | `Address` | [Address](#address)  | optional | Address of the customer. |

Reviewer: If we're calling it bill address, then it should be on both places.

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `Id` | string  | optional | ID of the [Customer](customers.md#customer) to whom the bill was assigned. |

Reviewer 2: is it correct to use "to whom the bill was assigned" here, or should it refer to the owner of the bill?

Author: At the moment, this refers to a customer to whom the bill was billed. In other words a customer's bill. It's not interchangeable with company.

| `Address` | [Bill address](#bill-address)  | optional | Address of the customer. |

> | `LegalIdentifiers` | Dictionary<string, string>  | optional | Legal Identifiers of the customer. |

Reviewer 2: I  assume type for `LegalIdentifiers` should be a new object containing a set of key:value pairs. If so, this needs to be defined with another table.

> | `BillingCode` | string  | optional | ID of the Customer to whom the bill was assigned. |

Reviewer 2: Again, is it okay to use "assigned"?

| `LastName` | string  | optional | Last name of the customer. |
| `FirstName` | string  | optional | First name of the customer. |
| `SecondLastName` | string  | optional | Second last name of the customer. |
| `TitlePrefix` | [Title](customers.md#Title)  | optional | Title prefix of the customer. |

#### Bill company data

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |

> | `Id` | string  | optional | ID of the [Company](companies.md#company) to whom the bill was assigned. |

Reviewer 2: ```suggestion
| `Id` | string  | optional | ID of the [Company](companies.md#company). |
```

| `Address` | [Bill address](#bill-address)  | optional | Address of the company. |

> | `LegalIdentifiers` | Dictionary<string, string>  | optional | Legal Identifiers of the company. |

Reviewer 2: I  assume type for `LegalIdentifiers` should be a new object containing a set of key:value pairs. If so, this needs to be defined with another table.

> | `BillingCode` | string  | optional | ID of the company to whom the bill was assigned. |

Reviewer 2: This can't be right, there is already a field for the company ID. What is BillingCode?

Author: Taken from our documentation. 'A unique code for Mews to list on invoices it sends to the company. The company can then easily recognize that the invoices are from your property.' I'll amend it to that, without the second sentence.

> | `Name` | string  | optional | Name of the company to which the bill was assigned. |

Reviewer 2: ```suggestion
| `Name` | string  | optional | Name of the company. |
```

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `CustomerID` | string  | optional | ID of the Customer to whom the bill was assigned. |
| `Address` | [Address](#address)  | optional | Address of the customer. |
| `LegalIdentifiers` | Dictionary<string, string>  | optional | Legal Identifiers of the customer. |
| `BillingCode` | string  | optional | ID of the Customer to whom the bill was assigned. |
| `LastName` | string  | optional | Last name of the customer. |
| `FirstName` | string  | optional | First name of the customer. |
| `SecondLastName` | string  | optional | Second last name of the customer. |
| `TitlePrefix` | string  | optional | Title prefix of the customer. |

> | `ItalianFiscalCode` | string  | deprecated | Italian fiscal code. Deprecated. Use values from LegalIdentifiers |

Reviewer 2: ```suggestion
| ~~`ItalianFiscalCode`~~ | ~~string~~ | ~~optional~~ | ~~Italian fiscal code.~~ Deprecated. Use values from `LegalIdentifiers` |
```

Reviewer 2: Any deprecations such as this should be listed in the table on the [deprecations](../deprecations/README.md) page, so please update that page as well.

Reviewer: @MikeAdamsMews I thought that if we deprecate it, we remove it from this table completely?

Reviewer 2: @KaliCZ please confirm with @vitezp. We discussed changing the policy so that we use strikethrough for all deprecations. Regardless, they should be listed on the deprecations page.

> | `ItalianLotteryCode` | string  | deprecated | Italian lottery code. Deprecated. Use values from LegalIdentifiers |

Reviewer 2: This is shown as deprecated, yet it was not in the previous version of documentation. Has it been agreed to add historic deprecations to the documentation? If so, please use the same format as suggested for `ItalianFiscalCode`, i.e. using ~~ strikethrough~~ . If not, it shouldn't be here.

> #### Bill company data

Reviewer 2: Is this the same 'Bill company data' that is referenced from both AssigneeData.Value and OptionalCompanyData? It seems confusing.

Author: Yes, the AssigneeData.Value and OptionalCompanyData share the same object definition, though they can have different values

| `Iban` | string  | optional | Enterprise IBAN (International Bank Account Number). |
| `Bic` | string  | optional | Enterprise BIC (Bank Identifier Code). |

> #### Bill address

Reviewer 2: ```suggestion
### Bill address
```

| `Iban` | string  | optional | Enterprise IBAN (International Bank Account Number). |
| `Bic` | string  | optional | Enterprise BIC (Bank Identifier Code). |

### Bill address

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Line1` | string  | optional | First line of the address. |
| `Line2` | string  | optional | Second line of the address. |
| `City` | string  | optional | City of the address. |
| `Postal Code` | string  | optional | Postal code of the address. |
| `SubdivisionCode` | string  | optional | ISO 3166-2 code of the administrative division. |

> | `CountryCode` | string  | optional | ISO 3166-1 code of the country. |

Reviewer 3: Just double checking the `contract` column here. If the address will be provided, all of them can be null or not?

Author: Well you need to fill in at least one address field, so basically all of them are optional

#### Legal Identifiers

> A key value pair where key determines type of legal identifier and value is the string value of that identifier

Reviewer 3: ```suggestion
A key value pair where key determines type of legal identifier and value is the string value of that identifier.
```

* `TaxIdentifier`
* `CompanyName`
* `Siret`
* `Siren`
* `NafCode`
* `RcsCode`
* `LegalStatus`
* `RegisteredCapital`
* `CityOfRegistration`
* `TradesDirectoryRegistrationNumber`
* `ItDestinationCode`
* `ItFiscalCode`
* `ItLotteryCode`
* `HungarianVatCode`

> * `HungarianCompanyName`
> * `HungarianTaxPayerIdentifier`
> 

Reviewer 3: Is this enum finite or not? If it can change in the future, it should be presented here like this.
```suggestion
* `HungarianCompanyName`
* `HungarianTaxPayerIdentifier`
* ...
```

| `Iban` | string  | optional | Enterprise IBAN (International Bank Account Number). |
| `Bic` | string  | optional | Enterprise BIC (Bank Identifier Code). |

### Bill address

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| `Line1` | string  | optional | First line of the address. |
| `Line2` | string  | optional | Second line of the address. |
| `City` | string  | optional | City of the address. |
| `Postal Code` | string  | optional | Postal code of the address. |
| `SubdivisionCode` | string  | optional | ISO 3166-2 code of the administrative division. |
| `CountryCode` | string  | optional | ISO 3166-1 code of the country. |

> #### Legal Identifiers

Reviewer 4: @MikeAdamsMews FYI: Another example of the dictionary, key value pair.

Reviewer 2: Ok, here's my proposal:

### Dictionary

Dictionary is a collection of key-value pairs.

| Property | Type | Contract | Description |
| :-- | :-- | :-- | :-- |
| ?Key? | string | optional | Some value corresponding to the ?Key? unique identifier. Cannot be null. |

### Legal Identifiers

`LegalIdentifiers` is a collection of key-value pairs, where the key is the type of legal identifier and the value is the corresponding value of that identifier. Keys are as follows:

* `TaxIdentifier`
* `CompanyName`
* `Siret`
* `Siren`
* `NafCode`
* `RcsCode`
* `LegalStatus`
* `RegisteredCapital`
* `CityOfRegistration`
* `TradesDirectoryRegistrationNumber`
* `ItDestinationCode`
* `ItFiscalCode`
* `ItLotteryCode`
* `HungarianVatCode`
* `HungarianCompanyName`
* `HungarianTaxPayerIdentifier`
* ...

Reviewer 2: I will update language texts to also use Dictionary, and Dictionary will become a 'common object definition' and moved out of here to the new page of common object definitions, but this will be on a separate Pull Request. For the moment, it can be defined here.

Reviewer 5: - I would remove Company name as we have already PR prepared that will remove it from Legal identifiers.
- Also values of this dictionary can be empty, for example `TradesDirectoryRegistrationNumber` is optional.

Author: @sianovd good idea. Removed the Company Name.
@MikeAdamsMews I've amended the text a bit to also reference to the new definition of Dictionary

| `Iban` | string  | optional | Enterprise IBAN (International Bank Account Number). |
| `Bic` | string  | optional | Enterprise BIC (Bank Identifier Code). |

> #### Address

Reviewer: @MikeAdamsMews How do we document addresses? Surely they're all the same - using the GeneralAPI dto.

Reviewer 2: @KaliCZ the plan is to introduce a new page for 'common object definitions' and it sounds like Address would be a candidate for that. It's ready to go but just waiting for the right opportunity - there is too much activity on the docs at the moment! For now, you can just reference Address wherever it is defined in the docs, e.g. configuration.md#address (assuming of course that this is the same object definition).

Author: I have dug around and unfortunately I haven't found any suitable existing dto that would suffice here. I have found an Address defined in configuration.md and also an Address defined in addresses.md. Neither of them match the same object definition
